### ISSUE TITLE
Separate cross-compiler and native compiler options

### ISSUE THREAD

**Issue by nirbheek** [mesonbuild/meson#1040]
Currently we only set the cross-compiler options when cross-compiling and pass the same options to the native compiler too. This means we can't set different options for both, and this even broke in #1029.

We should have a way to set native compiler options separately from the cross options. I propose that we maintain separate cross and native options, and by default both native and cross compiler options should be derived from the `project` invocation. However, we should allow people to override it for the cross-compiler in the cross-info file.

This also makes sense as the overall structure since the cross-info file is supposed to contain all information necessary for cross-compilation.

**Comment by jpakkane** [comment]
I have thought about this and I really don't want separate options for both native and cross whenever possible. This duplicates _all_ options and we already have too many of them. We should have all the options but only once. So, for example, `c_std` would be settable but it should be the same for cross and host builds. Conversely `cpp_debugstl` should be settable but would only have an effect on that compiler (cross or native) that supports it.

This might not be feasible in the end, but option explosion is something we should avoid whenever possible.

**Comment by nirbheek** [comment]
I didn't mean that we add separate options for cross-compilation. I mean that the `c_*` options in `meson.build` should apply to both native and cross-compilation by default, and the cross-info file should be able to override, for instance, `c_std`. There is no option explosion this way.

**Comment by Ericson2314** [comment]
This is done, and done symmetrically, internally. #5263 will expose that symmetrical implementation to the user.

I original had `build_c_*` `host_c_*` and `c_*` overriding both as you say @nirbheek, but then @jpakkane and @dcbaker peferred just `build_c_*`  `c_*` so I did that. The option explosion is "legit" in that in native and cross builds alike `native: true` and `native: false` can have separate args. While this may seem like overkill in the native case, @dcbaker and I argued that it's very good when native and cross can do the exact same thing in order to let users of meson rule out bugs and worry about fewer total combinations to test.
